contrib-compiler.md

```diff
@@ -0,0 +1,99 @@
+
+
+Thanks for interest in contributions.
+
+Note currently our philosphy is to make release build as easy as possible, while dev build is fairly complex at this time.
+
+## Set up dev dependencies
+
+- Having [https://opam.ocaml.org/](opam) installed
+
+    opam switch 4.02.3+buckle-master # use our OCaml compiler
+    opam install camlp4 cppo  <1>
+
+   Both `Camlp4` and `cppo` are used to generate OCaml code for pre-processing. 
+   Switch to our compiler makes your life easier (avoid confusion with diferent compiler versions)
+- Having [https://nodejs.org/](NodeJS) installed
+- Having GNU Make installed
+- OS: Mac/Linux (Note BuckleScript works well on Windows, but the dev mode is not tested)
+- Our dev build depends on a submodule
+
+  ```sh
+  git submodule --update
+  ``` 
+
+All source code are in `jscomp` directory, below assume that you are working in `jscomp` directory.
+
+
+## components of the compiler
+
+`bs-platform` is composed of several components, the core compiler, the bundled ocaml stdlib,  shipped js bindings, test cases and some internal tools.
+
+### internal tools
+
+- bspack
+
+  bspack is heavily used in `bs-platform`, it is a tool to pack all ocaml 
+  source code  into a single file, so that in release mode, there is not any dev dependency anymore.
+
+  It is very stable and heavily tested, in most cases, contributors don't need change its behavior.
```

Comment:
Might be fun to reference the issue for pulling it out into a separate library https://github.com/bloomberg/bucklescript/issues/1254, would be an easy thing for someone who is used to submitting OPAM packages